### PR TITLE
Fix/proof generation

### LINKED ISSUE #7071: Unable to serve snap data to Geth
My setup is:
1. Synced Nethermind nodes (one is 1.26.0 and one is master)
2. Geth node connects to one of them only

On master I got exception on geth side:
`DEBUG[05-24|09:13:18.194] Message handling failed in `snap`        peer=1b9b8aac err="proof node (hash 479df943983f349248bcd24a5aef400211ed335a43a8d1338799d016b6c75b14) missing"`

On 1.26.0 I got:
```
DEBUG[05-24|10:10:28.412] Message handling failed in `snap`        peer=d2ee026f err="invalid message: message msg #3 (12 bytes): invalid message: (code 3) (size 12) rlp: value size exceeds available input length"
TRACE[05-24|10:10:28.412] Protocol snap/1 failed                   id=d2ee026fe19b36a7 conn=staticdial err="invalid message: message msg #3 (12 bytes): invalid message: (code 3) (size 12) rlp: value size exceeds available input length"
```

### PR DESCRIPTION
**Pull request by asdacap**
Fix #7071 

- The assumption that the proof would have its key equal to the prefix of the first leaf is only true if the first leaf is close to the start hash or equal to start hash.
- Geth split ranges of storage trees which mean the start hash can be far from the first leaf. Nethermind does not, so this does not affect nethermind accept for top level tree, which for mainnet is large making this issue unlikely to show up.
- This changes the proof tracking logic to use the type of proof node to know what is the correct level for the next proof node.
- Also fix right proof randomly broken when cancelled.

#### What types of changes does your code introduce?

- [X] Bugfix (a non-breaking change that fixes an issue)

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Notes on testing

- [X] Geth can complete snap and healing from nethermind.
- [X] Nethermind can complete snap and healing from nethermind.
- [X] Snap hive test passed.

### CODE REVIEW

File: src/Nethermind/Nethermind.Synchronization.Test/RangeQueryVisitorTests.cs [hunk 1]
```diff
@@ -136,6 +138,64 @@ public void RangeFetchPartialLimit()
...
+        var startHash = new Hash256("0x1140000000000000000000000000000000000000000000000000000000000000");
+        var limitHash = new Hash256("0x1235000000000000000000000000000000000000000000000000000000000000");
+
+        RlpCollector leafCollector = new();
+        using RangeQueryVisitor visitor = new(startHash, limitHash, leafCollector);
+        stateTree.Accept(visitor, stateTree.RootHash, CreateVisitingOptions());
+
+        leafCollector.Leafs.Count.Should().Be(4);
+
+        ArrayPoolList<byte[]> proofs = visitor.GetProofs();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization.Test/RangeQueryVisitorTests.cs:177]: missed `using`

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs [hunk 2]
```diff
@@ -104,20 +104,18 @@ private bool IsRootMissing(in ValueHash256 stateRoot)
...
                     try
                     {
-                        Hash256 storagePath = new Hash256(requestedPath[0]);
+                        Hash256 storagePath = new Hash256(ValueKeccak.Zero);
+                        requestedPath[0].CopyTo(storagePath.Bytes); // One of the hive tests does not have complete 32 byte for storage path for some reason.
                         Account? account = GetAccountByPath(tree, rootHash, requestedPath[0]);
                         if (account is not null)
                         {
                             Hash256? storageRoot = account.StorageRoot;
-                            if (!storageRoot.Bytes.SequenceEqual(Keccak.EmptyTreeHash.Bytes))
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs:112]: This `if` is not helpful?

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs [hunk 3]
```diff
@@ -104,20 +104,18 @@ private bool IsRootMissing(in ValueHash256 stateRoot)
                 default:
                     try
                     {
-                        Hash256 storagePath = new Hash256(requestedPath[0]);
+                        Hash256 storagePath = new Hash256(ValueKeccak.Zero);
+                        requestedPath[0].CopyTo(storagePath.Bytes); // One of the hive tests does not have complete 32 byte for storage path for some reason.
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs:108]: Avoid double copying?

```suggestion
                        Hash256 storagePath = new Hash256(requestedPath[0].Length == Hash256.Size ? requestedPath[0] : requestedPath[0].PadRight(Hash256.Size));
```

File: src/Nethermind/Nethermind.Trie/RangeQueryVisitor.cs [hunk 4]
```diff
@@ -126,41 +125,31 @@ public long GetBytesSize()
...
             {
-                if (!_rightmostNodes[i].HasValue) continue;
-
-                (TreePath path, TrieNode node) = _rightmostNodes[i].Value;
-                rightmostPath.TruncateMut(i);
-                if (rightmostPath != path) continue;
+                TrieNode node = boundaryNodes[i];
+                if (node is null) break;
 
                 proofs.Add(node.FullRlp.ToArray());
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/RangeQueryVisitor.cs:143]: maybe make it `ArrayPoolList<ArrayPoolList<byte>>`?

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs:112]: Ow.. yea, it failed hive test.

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**